### PR TITLE
HIVE-25814: Add entry in replication_metrics table for skipped replication.

### PR DESCRIPTION
**Pull request by hmangla98**

### CODE REVIEW

File: common/src/java/org/apache/hadoop/hive/ql/ErrorMsg.java [hunk 1]
```diff
@@ -614,7 +614,6 @@
 
   SPARK_JOB_RUNTIME_ERROR(40001, "Spark job failed due to: {0}", true),
   SPARK_TASK_RUNTIME_ERROR(40002, "Spark job failed due to task failures: {0}", true),
-  REPL_DATABASE_IS_TARGET_OF_REPLICATION(40003, "Cannot dump database as it is a Target of replication."),
```
Comment (ayushtkn) [inline common/src/java/org/apache/hadoop/hive/ql/ErrorMsg.java:617]: Unrelated to the change, we are doing here, can avoid this

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/util/ReplUtils.java [hunk 2]
```diff
@@ -474,4 +475,17 @@ public static void addLoggerTask(ReplLogger replLogger, List<Task<?>> tasks, Hiv
       DAGTraversal.traverse(tasks, new AddDependencyToLeaves(Collections.singletonList(task)));
     }
   }
+
+  public static void reportStatusInReplicationMetrics(String stageName, Status status, String errorLogPath,
+                                                      HiveConf conf)
+          throws SemanticException {
+    GenericMetricCollector metricCollector = new GenericMetricCollector(0, conf);
```
Comment (ayushtkn) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/util/ReplUtils.java:482]: What advantage is ``GenericMetricCollector`` class giving you?
It is just calling the super constructor with pre-set value, that you could do it here itself. Or by having a Constructor in the parent class which takes only two params and set other values as ``null``

**Review by ayushtkn — COMMENTED** [review]
This doesn't seems to cover the load iterations skipped? 
A load iteration can also be skipped, if there is no dump or if the previous load is in progress or load failed with some error.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/util/ReplUtils.java (hunk 2 above)
Comment (hmangla98) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/util/ReplUtils.java:482]: Done

**Comment by hmangla98** [comment]
> This doesn't seems to cover the load iterations skipped? A load iteration can also be skipped, if there is no dump or if the previous load is in progress or load failed with some error.

I've covered that case in this change only. Can you please go through TestReplicationScenariosAcidTables#testReplicationMetricForSkippedIteration?

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcidTables.java [hunk 3]
```diff
@@ -156,6 +160,106 @@ public void tearDown() throws Throwable {
...
+    metric = collector.getMetrics().getLast();
+    assertEquals(metric.getProgress().getStatus(), Status.FAILED_ADMIN);
+    assertEquals(metric.getProgress().getStages().get(0).getErrorLogPath(), nonRecoverableFile.toString());
+
+    replica.loadFailure(replicatedDbName, primaryDbName);
+
+    metric = collector.getMetrics().getLast();
+    assertEquals(metric.getProgress().getStatus(), Status.SKIPPED);
+    assertEquals(metric.getProgress().getStages().get(0).getErrorLogPath(), nonRecoverableFile.toString());
+    isMetricsEnabledForTests(false);
```
Comment (ayushtkn) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcidTables.java:260]: move this to finally block

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenarios.java [hunk 4]
```diff
@@ -4610,6 +4613,99 @@ public void testPolicyIdImplicitly() throws Exception {
...
+  @Test
+  public void testReplicationMetricForSkippedIteration() throws Throwable {
+    String name = testName.getMethodName();
+    String primaryDbName = createDB(name, driver);
+    String replicaDbName = "replicaDb";
+    try {
+      isMetricsEnabledForTests(true);
+      MetricCollector collector = MetricCollector.getInstance();
+      run("create table " + primaryDbName + ".t1 (id int) clustered by(id) into 3 buckets stored as orc tblproperties (\"transactional\"=\"true\")", driver);
+      run("insert into " + primaryDbName + ".t1 values(1)", driver);
```
Comment (ayushtkn) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenarios.java:4625]: I think the insert is failing due to transactional true.
I tried a select * on this
```
java.io.IOException: java.lang.RuntimeException: ORC split generation failed with exception: java.lang.IllegalArgumentException: No ValidTxnList for pfile:/Users/ayushsaxena/code/hive-os/hive/itests/hive-unit/target/warehouse/testreplicationmetricforskippediteration.db/t1/delta_0000000_0000000_0000/000001_0/delta_0000000_0000000

```

Put this after insert and check:
```
      verifySetupSteps = true;
      verifySetup("SELECT id from " + primaryDbName + ".t1", new String[]{"1"}, driver);
```

**Review by ayushtkn — APPROVED** [review]
LGTM

**Pull request merged**